### PR TITLE
IOGX Integration [Work In Progress]

### PR DESCRIPTION
**Pull request by zeme-iohk**
As part of the ongoing effort to build a Developer Platform for IOG, the smart-contracts-dev-empowerment team is releasing a flake for a standard development environment + CI for Haskell-based projects. 
We wish to use `marlowe-cardano` as the pilot project to test this.

This PR adds the `__iogx__` folder to the top-level, which contains the new nix code.
Running `cloc` will tell you that the new nix code is ~50% shorter than the old nix code (because the functionality shared by the smart-contracts repos has been moved to the new `github:zeme-iohk/iogx` flake).

This PR merges the new flake into the existing one. All new flake outputs are nested inside the attribute `__iogx__` (e.g. the new packages are in `packages.__iogx__`
This includes `hydraJobs`, so that both CIs can be tested in parallel.
 
If you wish to test this, all you have to do is run `nix develop .#__iogx__.default` instead of `nix develop`.
Note that both the old and new nix code can coexist, so this PR can be merged now without breaking anything.
If the new `hydraJobs` fail, PRs can still be merged because the old `required` job isn't affected by this.

If this PR is accepted, you are invited to start using new shell (`nix develop .#__iogx__.default`) only, and if you encounter any problem with this new shell, or if you want to suggest any improvements, ping @zeme-iohk on GitHub or @zeme on Slack, and then you can go back to using the old `nix develop`.

To complete migration, all the old nix code can be deleted, and the `__iogx__` folder can be renamed to `nix`. 

The `bitte` and `deploy` folders have been moved to `__iogx__` and the contents are largely unchanged.
The remaining nix files have either been deleted or refactored beyond recognition. 
To review you should start from `__iogx__/flake.nix`

### CODE REVIEW

File: __iogx__/marlowe-cardano/bitte/node-socat.nix [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{ writeShellScriptBin, wait-for-socket, socat, lib, coreutils }:
```
Comment (zeme-iohk) [inline __iogx__/marlowe-cardano/bitte/node-socat.nix:1]: This file is practically identical to the original.

File: __iogx__/marlowe-cardano/bitte/node/default.nix [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# Needed variables:
```
Comment (zeme-iohk) [inline __iogx__/marlowe-cardano/bitte/node/default.nix:1]: This file is practically identical to the original.

File: __iogx__/marlowe-cardano/deploy/README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# Developer Service Deployment
```
Comment (zeme-iohk) [inline __iogx__/marlowe-cardano/deploy/README.md:1]: This file has just been copy-pasted

File: __iogx__/marlowe-cardano/deploy/nomadTasks.nix [hunk 4]
```diff
@@ -0,0 +1,188 @@
+{ inputs }:
```
Comment (zeme-iohk) [inline __iogx__/marlowe-cardano/deploy/nomadTasks.nix:1]: This file is practically identical to the original.

File: __iogx__/marlowe-cardano/deploy/oci-images.nix [hunk 5]
```diff
@@ -0,0 +1,63 @@
+{ inputs, pkgs }:
```
Comment (zeme-iohk) [inline __iogx__/marlowe-cardano/deploy/oci-images.nix:1]: This file is practically identical to the original.

File: __iogx__/marlowe-cardano/scripts.nix [hunk 6]
```diff
@@ -0,0 +1,76 @@
+{ inputs, pkgs }:
```
Comment (zeme-iohk) [inline __iogx__/marlowe-cardano/scripts.nix:1]: This file is practically identical to the original.

File: __iogx__/per-system-outputs.nix [hunk 7]
```diff
@@ -0,0 +1,50 @@
+# The actual flake outputs, per system.
```
Comment (zeme-iohk) [inline __iogx__/per-system-outputs.nix:1]: Marlowe-cardano _specific_ flake outputs. All haskell+devenv stuff is handled by the underlying iogx flake

File: __iogx__/tullia.nix [hunk 8]
```diff
@@ -0,0 +1,73 @@
+/*
```
Comment (zeme-iohk) [inline __iogx__/tullia.nix:1]: This file is practically identical to the original.

File: flake.nix [hunk 9]
```diff
@@ -56,153 +56,175 @@
...
-                system = system';
-                inherit enableHaskellProfiling source-repo-override inputs crossSystem evalSystem;
-                inherit (inputs) haskell-nix;
-              };
-            })
-            systems);
-        };
-    };
+    in
+    inputs.nixpkgs.lib.recursiveUpdate originalFlake newFlake;
```
Comment (zeme-iohk) [inline flake.nix:219]: Here's where we merge the two flakes together (old and new)

File: hydra-jobs.nix [hunk 10]
```diff
@@ -54,6 +54,6 @@ let
   ciJobsets = stripAttrsForHydra (filterDerivations ci);
 in
 traceNames "" (ciJobsets // {
-  required = derivationAggregate "required-marlowe" ciJobsets;
+  required = derivationAggregate "required-marlowe" { ${evalSystem} = ciJobsets; };
```
Comment (zeme-iohk) [inline hydra-jobs.nix:57]: Needed this to be able to have both hydra jobsets alive together

File: hydra-jobs.nix [hunk 11]
```diff
@@ -54,6 +54,5 @@ let
   ciJobsets = stripAttrsForHydra (filterDerivations ci);
 in
 traceNames "" (ciJobsets // {
-  required = derivationAggregate "required-marlowe" ciJobsets;
```
Comment (zeme-iohk) [inline hydra-jobs.nix:57]: A little magic needed in order to merge the new hydraJobs with the old hydraJobs.
The `required` job is defined in flake.nix now

File: .devenv/profile [hunk 12]
```diff
@@ -0,0 +1 @@
+/nix/store/xv832r4rvmcsxp4z6kxvchs0qgwl5czd-devenv-profile
```
Comment (jhbertra) [inline .devenv/profile:1]: Should this be committed?

File: __iogx__/per-system-outputs.nix [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  nomadTasks = import ./marlowe-cardano/deploy/nomadTasks.nix
+    { inherit inputs; };
+
+  networks = import ./marlowe-cardano/networks.nix
+    { inherit inputs pkgs; };
+
+  packages.entrypoints = import ./marlowe-cardano/bitte
+    { inherit inputs pkgs; };
+
+} // inputs.tullia.fromSimple pkgs.stdenv.system (import ./tullia.nix)
```
Comment (jhbertra) [inline __iogx__/per-system-outputs.nix:50]: Feel free to get rid of this (It's not used)

File: .devenv/profile (hunk 12 above)
Comment (zeme-iohk) [reply .devenv/profile:1]: No, good catch, I will remove

File: __iogx__/marlowe-cardano/deploy/operables.nix [hunk 14]
```diff
@@ -0,0 +1,538 @@
+{ inputs, pkgs }:
```
Comment (zeme-iohk) [inline __iogx__/marlowe-cardano/deploy/operables.nix:1]: This file is practically identical to the original

**Comment by jhbertra** [comment]
I entered the nix shell - nice MOD (some descriptions need work, and there may be some missing commands, but it's nice to see everything the shell provides).

I did encounter the following issue running `cabal` though:

```bash
$ cabal
/nix/store/if0kap3xxbnnp7kk9jy84zgd6jz8dajw-cabal/bin/cabal: line 2: /nix/store/jv8mha93hdfkwigmzxxgwa8rnwyn0ir0-cabal-install-exe-cabal-3.6.2.0: Is a directory
```

**Comment by jhbertra** [comment]
There are two commands missing from the shell:

```nix
  marlowe-runtime-cli = mkCabalExeScript "marlowe-runtime-cli" "marlowe-runtime-cli";
  marlowe-cli = mkCabalExeScript "marlowe-cli" "marlowe-cli";
```

**Comment by shlevy** [comment]
@zeme-wana Is this suitable for review again?

**Comment by zeme-wana** [comment]
> @zeme-wana Is this suitable for review again?

Yes it can be reviewed now

**Review by shlevy — APPROVED** [review]
LGTM with brief skim

**Review by jhbertra — APPROVED** [review]
LGTM!

**Pull request merged**